proposals/LayerProtocolRevision.md

```diff
@@ -0,0 +1,172 @@
+
+    let gradients = gradient(at: model) { model -> Tensor<Float> in
+      let logits = model.applied(to: images)
+      let batchLoss = softmaxCrossEntropy(logits: logits, labels: labels)
+      totalLoss += batchLoss.scalarized()
+      return batchLoss
+    }
+    optimizer.update(&model.allDifferentiableVariables, along: gradients)
+  }
+  // test
+  testQueue.async(group: testGroup) {
+    var totalCorrect: Int32 = 0
+    for i in 0..<Int32(10) {
+      let images = minibatch(testImages, size: testBatchSize, batchIndex: i)
+      let labels = minibatch(numericTestLabels, size: testBatchSize, batchIndex: i)
+      let predictions = model.infer(from: images)
+      let correct = predictions.argmax(squeezingAxis: 1) .== labels
+      totalCorrect += Tensor<Int32>(correct).sum().scalarized()
+    }
+
+    let accuracy = Float(totalCorrect) / Float(numericTestLabels.shape[0])
+    print("epoch \(epoch) accuracy: \(accuracy) loss: \(totalLoss)")
+  }
+}
+testGroup.wait()
+print("Training complete: \(String(timeInterval: Date().timeIntervalSince(start)))")
+```
+## Copying Layers and the LearningPhaseIndicator
+Some operators such as BatchNorm and Dropout need to behave differently depending on whether they are performing training or inference. The current design defines the LearningPhaseIndicator class which is intended to behave like a global variable scoped to a single model. The training loop would toggle the _.training_ value depending on whether training or inference is being performed.
+
+The examples I saw in [tensorflow swift-models](https://github.com/tensorflow/swift-models) had the LearningPhaseIndicator declared and manipulated separately from the model it was affecting. One object having a side effect on another is problematic. Declaring it as a member of the root layer would have been better. In any case this design won’t work because as soon as you copy a model, the same LearningPhaseIndicator will be affecting both models. This would make it impossible to perform concurrent testing, or work with model copies in general. I don’t believe there is any clean way to have a pseudo global variable scoped to a Layer tree. I ran into the same design problem several years ago.
+
+## Suggested Design Change
+A simple and performative solution is to modify the Layer _applied(to:_ API to include a training parameter. Perhaps add an extension function to drop the parameter when calling layers that don’t make a distinction.
+```swift
+public protocol Layer: Differentiable & KeyPathIterable
+    where AllDifferentiableVariables: KeyPathIterable {
+    /// The input type of the layer.
+    associatedtype Input: Differentiable
+    /// The output type of the layer.
+    associatedtype Output: Differentiable
+
+    /// Returns the output obtained from applying to an input.
+    @differentiable(wrt: (self, input))
+    func applied(to input: Input, training: Bool) -> Output
```

Comment:
```suggestion
    func applied(to input: Input, in context: Context) -> Output
```

where `Context` is defined as (something similar to):

```swift
protocol Context {
  var training: Bool { get }
}
```

The reason being is that: I think it'd be a good idea to make this a bit more general / flexible. I suspect that it may be valuable for layers to (eventually) be able to know a bit more about the context within which they're executing. e.g. if we're fine-tuning for quantization. This way, if we decide we need to have an additional orthogonal value (or if we decide that there's actually more than 2 modes (e.g. not just training and inference but say fine-tuning) we have the flexibility to make that change without causing a lot of churn.

The risk I see with going in this direction, however, is that the `Context` object could easily become a dumping ground for arbitrary utility things. So, it'd be very good we are careful.

Additionally, I think there's an open question as to whether `context` should be the first parameter or the second parameter. (e.g. Golang as a similar pattern, and always requires `Context` to be the first value.)

Finally, I've suggested `Context` to be a protocol, but I'm not sure if it would better be served as a `struct`. I imagine this decision has some flexibility tradeoffs, as well as performance implications. I'd be curious to hear more from those who have carefully tuned performance in Swift.

What do folks think of this direction?

Reply:
I think defining a Context protocol opens the door to forward flexibility and is a reasonable idea. If the notion of a Context is adopted then I agree stylistically it would be best as the first parameter.

Models are intended to run in a variety of execution contexts, so this could be an ideal way to bundle up and pass through a collection of control parameters from User level to the back end. For example there are a variety of parameters that can be used to specify Cuda behavior, and in the GCP case there might be a completely different set of context information for accounts, specific resources, and resource configuration parameters.

A set of objects conforming to _Context_ with additional specialized protocols for each of the common environments (CPU, Cuda, GCP) could be included as part of the product to cover 99% of the user scenarios. The collection could be added to over time if needed and passed through without requiring modifications to the rest of the API.

The compiler seems to do a pretty good job of efficiently passing structs down a call tree. I believe even though they are value types, the compiler simply passes a ref until there is an attempt to mutate, which I don't expect would happen other than at initialization time.

Your concern about the context being abused and becoming a dumping ground for poorly designed 3rd party components is quite valid. I don't know what the company policy is about asserting control over things. I expect most users will have no need to use anything other than the standard set. In any case, clear guidelines should be provided to discuss the issues involved in creating a custom context.